sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,755 @@
+introduces a default memory pool per device for each SYCL context and device
+pair for device allocations and a default memory pool per context for host
+allocations.
+
+New member functions are added to the `context` class to retrieve the default
+memory pool as a copy of the `memory_pool` object. This can be modified and have
+those modifications reflected as it conforms to the SYCL common reference
+semantics.
+
+[source,c++]
+----
+class context {
+
+  memory_pool context::ext_oneapi_get_default_memory_pool() const;
+
+  memory_pool context::ext_oneapi_get_default_memory_pool(device dev) const;
+
+}; // context
+----
+
+[source, c++]
+----
+memory_pool context::ext_oneapi_get_default_memory_pool() const;
+----
+
+_Returns_: The default memory pool associated with the context for allocating
+with the allocation kind `usm::host`.
+
+[source, c++]
+----
+memory_pool context::ext_oneapi_get_default_memory_pool(device dev) const;
+----
+
+_Returns_: The default memory pool associated with the context and `dev` for
+allocating with the allocation kind `usm::device`.
+
+
+=== Asynchronous malloc & free
+
+This extension introduces a series of new enqueue functions for enqueueing
+asynchronous malloc and free commands which operate with the memory pools also
+introduced in this extension.
+
+All enqueue functions introduced have overloads which take a SYCL `queue` and a
+SYCL `handler`. None of enqueue functions return a SYCL `event` direction, as
+this extension is in line with the
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
+calling `submit_with_event` and the `handler` overloads of these enqueue
+functions.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+void *async_malloc(queue q, size_t size);
+
+void *async_malloc(handler h, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+
+void async_free(queue q, void *ptr);
+
+void async_free(handler h, void *ptr);
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+void *async_malloc(queue q, size_t size);
+
+void *async_malloc(handler h, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+----
+
+_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which
+will asynchronously allocate memory of size `size` in bytes, allocating from the
+memory pool `pool` if provided, otherwise allocation from the default memory
+pool associated with the SYCL context and device associated with `q` or `h`.
+Memory is first allocated from the memory pool if possible, otherwise memory is
+allocated from the source to the memory pool to provide enough memory in the
+memory pool for the allocation.
```

Comment:
I think this needs to be wording in the spec as this behaviour is what guarantees the benefit of using the memory pool, the "if possible" part does provide flexibility to an implementation to decide when this happens and when it doesn't.

Reply:
I'm still not convinced this should be in the spec, especially since we say "if possible".  If it's not a hard requirement, how can you test for it?  This sounds more like something that is left to the implementation as a "quality of implement" choice.